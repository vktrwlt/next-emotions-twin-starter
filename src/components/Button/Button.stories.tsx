import React from 'react';
import { Story } from '@storybook/react/types-6-0';
import { Button, ButtonProps } from '@/components/Button';

export default {
  title: 'Button',
  component: Button,
  argTypes: {
    variant: {
      description: 'The type of Button',
      type: 'inline-radio',
      control: {
        type: 'inline-radio',
        options: ['solid', 'ghost', 'outline', 'link'],
      },
    },
    size: {
      description: 'The size of Button',
      default: 'md',
      control: {
        type: 'inline-radio',
        options: ['xs', 'sm', 'md', 'lg'],
      },
    },
    children: {
      description: 'The Button content',
      defaultValue: 'Button',
      type: { name: 'text', required: true },
    },
  },
};

const Template: Story<ButtonProps> = (args) => <Button {...args} />;

export const Default = Template.bind({});
Default.args = { variant: 'solid', size: 'md' };
