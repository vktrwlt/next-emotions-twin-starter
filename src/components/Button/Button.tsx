import { FC } from 'react';
import tw, { styled, css, theme, TwStyle } from 'twin.macro';

export interface ButtonProps {
  /**
   * The visual style of the Button
   * @default "solid"
   */
  variant?: 'solid' | 'ghost' | 'outline' | 'link';
  /**
   * The size of the Button
   * @default "md"
   */
  size?: 'xs' | 'sm' | 'md' | 'lg';
  children: React.ReactNode;
}

const primaryVariants = {
  // Named class sets
  solid: tw`text-white bg-green-600 hocus:(bg-green-700)`,
  ghost: tw`text-green-700 border-green-700 border-2 hocus:(bg-green-100)`,
  outline: tw`text-green-700 hocus:(bg-green-100)`,
  link: tw`text-green-700 hocus:(underline)`,
};

const sizeVariants = {
  // Named class sets
  xs: tw`px-4 py-1`,
  sm: tw`px-5 py-2`,
  md: tw`px-6 py-3`,
  lg: tw`py-4 text-xl px-7`,
};

export const Button: FC<ButtonProps> = styled.button(({ variant, size }) => [
  // Base Styles
  tw`font-semibold duration-75 transform rounded focus:outline-none`,

  // Dynamic Styles
  primaryVariants[variant],
  sizeVariants[size],
]);

Button.defaultProps = {
  variant: 'solid',
  size: 'md',
};
