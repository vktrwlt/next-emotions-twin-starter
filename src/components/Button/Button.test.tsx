import { Button } from '@components/Button';
import { render, fireEvent } from '@utils/test-utils';

describe('<Button />', () => {
  const { asFragment } = render(<Button>Button</Button>);
  it('should render without crashing', () => {
    expect(asFragment()).toMatchSnapshot();
  });
});
