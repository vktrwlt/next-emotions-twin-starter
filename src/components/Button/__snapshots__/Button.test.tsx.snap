// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Button /> should render without crashing 1`] = `
<DocumentFragment>
  .emotion-0 {
  -webkit-transform: var(--tw-transform);
  -moz-transform: var(--tw-transform);
  -ms-transform: var(--tw-transform);
  transform: var(--tw-transform);
  font-weight: 600;
  transition-duration: 75ms;
  border-radius: 0.25rem;
  --tw-text-opacity: 1;
  color: rgba(255, 255, 255, var(--tw-text-opacity));
  --tw-bg-opacity: 1;
  background-color: rgba(5, 150, 105, var(--tw-bg-opacity));
  padding-left: 1.5rem;
  padding-right: 1.5rem;
  padding-top: 0.75rem;
  padding-bottom: 0.75rem;
}

.emotion-0:focus {
  outline: 2px solid transparent;
  outline-offset: 2px;
}

.emotion-0:hover,
.emotion-0:focus {
  --tw-bg-opacity: 1;
  background-color: rgba(4, 120, 87, var(--tw-bg-opacity));
}

<button
    class="emotion-0 emotion-1"
  >
    Button
  </button>
</DocumentFragment>
`;
